### ISSUE TITLE
Formatter changes the value of implicitly concatenated strings when joining after an octal escape

### ISSUE THREAD

**Issue by SerhMosk** [astral-sh/ruff#28842]
### Summary

When `ruff format` joins implicitly concatenated string or bytes literals into one literal, an octal escape at the end of one part can absorb leading octal digits of the next part. This changes the runtime value of the literal.

```python
# octal.py
x = b"A\0" b"1\0"
y = "\1" "2"
print(x, repr(y))
```

```console
$ python octal.py
b'A\x001\x00' '\x012'
$ ruff format --isolated octal.py
1 file reformatted
$ cat octal.py
x = b"A\01\0"
y = "\12"
print(x, repr(y))
$ python octal.py
b'A\x01\x00' '\n'
```

`b"A\0" b"1\0"` becomes `b"A\01\0"`, which is a different value. The same happens with `str` literals: `"\1" "2"` becomes `"\12"`, which is `"\n"`.

The trigger is a part that ends with a one- or two-digit octal escape (`\0`, `\1`, `\12`, …) followed by a part that starts with an octal digit `0`–`7`. CPython resolves the escapes in each part before it concatenates them, so the joined text is parsed differently.

The linter had the same problem in the ISC001 fix, which #12936 fixed. The formatter's joining of implicitly concatenated strings seems to still have it.

Expected behavior: either leave such parts unjoined, or pad the escape to three digits when joining (`b"A\0001\0"`, `"\0012"`). Both keep the value.

We found this while checking that a one-off `ruff format` pass over a codebase changed formatting only. It showed up because we compared the AST before and after the pass.

### Version

ruff 0.15.18 (6686f6340 2026-06-18); also reproduced with ruff 0.16.8

**Comment by ntBre** [comment]
Thank you for the report!
